src/Reflection/EntityFieldReflection.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace PHPStan\Reflection;
+
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+/**
+ * Allows field access via magic methods
+ *
+ * See \Drupal\Core\Entity\ContentEntityBase::__get and ::__set.
+ */
+class EntityFieldReflection implements PropertyReflection
+{
+
+  /** @var ClassReflection */
+    private $declaringClass;
+
+  /** @var string */
+    private $propertyName;
+
+    public function __construct(ClassReflection $declaringClass, string $propertyName)
+    {
+        $this->declaringClass = $declaringClass;
+        $this->propertyName = $propertyName;
+    }
+
+    public function getType(): Type
+    {
+        if ($this->propertyName == 'original') {
+          // See Drupal\Core\Entity\EntityStorageBase::doPreSave
+            return new ObjectType('Drupal\Core\Field\FieldItemListInterface');
+        }
+
+        return new ObjectType('Drupal\Core\Field\FieldItemListInterface');
```

Comment:
This is tricky.  This applies only to ContentEntityBase

```
  /**
   * Implements the magic method for getting object properties.
   *
   * @todo: A lot of code still uses non-fields (e.g. $entity->content in view
   *   builders) by reference. Clean that up.
   */
  public function &__get($name) {
    // If this is an entity field, handle it accordingly. We first check whether
    // a field object has been already created. If not, we create one.
    if (isset($this->fields[$name][$this->activeLangcode])) {
      return $this->fields[$name][$this->activeLangcode];
    }
    // Inline getFieldDefinition() to speed things up.
    if (!isset($this->fieldDefinitions)) {
      $this->getFieldDefinitions();
    }
    if (isset($this->fieldDefinitions[$name])) {
      $return = $this->getTranslatedField($name, $this->activeLangcode);
      return $return;
    }
    // Else directly read/write plain values. That way, non-field entity
    // properties can always be accessed directly.
    if (!isset($this->values[$name])) {
      $this->values[$name] = NULL;
    }
    return $this->values[$name];
  }
```

It could be a FieldItemListInterface or regular property. I guess we can handle unique properties one by one like `original`.